[Tablet Orders] Update completion button based on flag, and navigation title based on size class

iamgabrielma opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #12057

## Description
This PR addresses 2 design tweaks:
* On iPhones, portrait: The product selector completion button currently displays the number of selected products when we select or deselect them. With the new designs this will only show either "Done" if no products have been selected, or "Add Products" text if any has been selected.
* On iPhones, portrait: The "selected products" text that was showing in the completion button, now has been moved to the navigation header.

<img width="282" alt="Screenshot 2024-02-28 at 11 08 40" src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/120993da-9f37-4f12-b744-f1da26591590">

## Testing instructions
* With `.sideBySideViewForOrderForm` disabled, observe that the view behaves as before: "Add Product" shows in the navigation bar, and the button is updated with the amount of selected products.
* With `.sideBySideViewForOrderForm` enabled, observe that the view behaves as after: The navigation bar shows the number of selected products, and the button only switches between "Done" (if no products have been selected) or "Add Products" (if any has been selected).

| iPhone before | iPhone after |
|--------|--------|
| ![simulator_screenshot_1EBEB86D-5A87-4031-87A5-0EEEDD373097](https://github.com/woocommerce/woocommerce-ios/assets/3812076/46a6b8e5-ed84-4cf0-a31c-356cf96b28ec) | ![simulator_screenshot_26EEC54D-6CFA-436F-A28C-B78A2A3E3F09](https://github.com/woocommerce/woocommerce-ios/assets/3812076/0d6d3732-7a86-4697-a255-345b98d726ba) | 

## Edge case in tablets:

There's one edge case that is not addressed in this PR: If the feature flag is enabled, and we run the app using split view where the product selector is narrow, but not narrow enough to use modal-on-modal, then we'll see only "Add Products" in the navigation title. This will not happen if we "move to that view", only when we initialize the product selector in that specific setup.

![20240228 edge case tablet splitview narrow](https://github.com/woocommerce/woocommerce-ios/assets/3812076/713bd2b3-497a-4b1a-a92d-65dce4edb128)

This is because the `onChange` property that checks the current subview width and needs to notify the parent in order to update the navigation title is not called on initialization, but only when the property changes. Addressing this seems to be a better fit when we attempt to switch from the current usage of GeometryReader to specific adaptive views.

iamgabrielma committed:
Update navigation title based on size class

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

iamgabrielma committed:
Update done button product selector view text

iamgabrielma committed:
Edge case for rendering navigation title on tablet

joshheald commented:
> This is because the `onChange` property that checks the current subview width and needs to notify the parent in order to update the navigation title is not called on initialization, but only when the property changes. Addressing this seems to be a better fit when we attempt to switch from the current usage of GeometryReader to specific adaptive views.

FYI using `onAppear` to do the same as `onChange` is a useful and common approach – you need both, but it might fix this.

joshheald reviewed (approved):
![No product selection count on iPad modal](https://github.com/woocommerce/woocommerce-ios/assets/2472348/eeef57b3-c509-4ada-b879-7c44dc6e8278)

I noticed that in a smaller iPad presentation, and on my iPhone however I get to it, both the title and the button say `Add Products`. This appears to be because the `isViewWidthNarrowerThanConstantRowWidth` property is never getting set.

![No count on iPhone](https://github.com/woocommerce/woocommerce-ios/assets/2472348/719cb6e9-b959-465b-a55c-2ed3135b338e)

Adding the `.onAppear` as I suggested in my last comment seems to fix this.

It'll need some more testing I think, when you've made that change... I just did a quick check.

joshheald commented on WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductSelectorView.swift:
```diff
@@ -79,9 +84,22 @@ struct ProductSelectorView: View {
+        guard ServiceLocator.featureFlagService.isFeatureFlagEnabled(.sideBySideViewForOrderForm) else {
+            return String.pluralize(viewModel.totalSelectedItemsCount,
+                                    singular: configuration.doneButtonTitleSingularFormat,
+                                    plural: configuration.doneButtonTitlePluralFormat)
```
can/should we use the `viewModel.selectProductsTitle` here too?

iamgabrielma replied:
Yes, when I was extracting logic outside of the view I saw we have some duplication when populating the view, some of it comes from the configuration and then we modify it within the view, however, these same string operations can be found in the view model. Changed on https://github.com/woocommerce/woocommerce-ios/pull/12126/commits/0383efbc764fdcde6abd25f58d0aa8842c5cb517

iamgabrielma committed:
Merge branch 'trunk' into issue/12057-product-count-to-navigation-header

# Conflicts:
#	WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductSelectorView.swift

iamgabrielma committed:
Update done button to use viewmodel data

There’s some duplication regarding data used to populate the view. Some data comes from the configuration file while some comes from the viewmodel.

iamgabrielma committed:
Add adjustViewWidthIfNeeded to .onAppear

iamgabrielma committed:
Check for size class on navigation title rendering

Checking the view’s width is not enough to render the exact titles proposed by design, we need to check as well if the size class is compact

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/188">17.6</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

iamgabrielma commented:
Thanks for the review!

> Adding the .onAppear as I suggested in my last comment seems to fix this. It'll need some more testing I think, when you've made that change... I just did a quick check.

That seemed to fix it initially, but I found more edge cases mainly related to using a constant of 450 units as threshold, which works fine on iPhone is not enough for some bigger views, as well as what seems to be the SwiftUI view lifecycle, since checking the view's width `.onAppear` returns 0.0 in some instances unless we perform a DispatchQueue there, which adds more complexity.

Along with your `.onAppear` suggestion, I added an additional check for the size class that seems to fix all cases I tested: https://github.com/woocommerce/woocommerce-ios/pull/12126/commits/439c9a5c44cd1376a9e121eb43c74661b82b146d
